client/src/main.rs

```diff
@@ -136,68 +194,124 @@ async fn main() -> Result<(), Box<dyn Error>> {
         .read_packet(&secret, nonce_generator_read.as_mut())
         .await;
     match p {
-        Ok(Some(ClientboundPacket::EncryptionAck)) => {
-            println!("Encryption handshake successful!");
-        }
+        Ok(Some(ClientboundPacket::EncryptionAck)) => {}
         Ok(_) => {
-            println!("Failed encryption step 2. Server response: {:?}", p);
-            std::process::exit(1);
+            panic!("Failed encryption step 2. Server response: {:?}", p);
         }
         Err(e) => {
-            println!("{}", e);
-            std::process::exit(1);
+            panic!("{}", e);
         }
     }
 
     //==================================
     //      Get credentials
     //==================================
-    let mut stdio = tokio::io::stdin();
-    let username = loop {
-        println!("Username:");
-        let mut buf = bytes::BytesMut::new();
-        match stdio.read_buf(&mut buf).await {
-            Ok(0 | 1) => println!("Username can't be empty!"),
-            Ok(l) => {
-                if l > 18 {
-                    println!("Username too long. (Max 17 characters)");
-                    continue;
-                }
-                let s = String::from_utf8_lossy(buf.strip_suffix(OS_EOL).unwrap()).to_string();
-                println!("{:?}", s);
-                if s.chars().any(|c| !c.is_alphanumeric()) {
-                    println!("Invalid characters in username.");
-                } else {
-                    break s;
+
+    while !w_login.is_finished() {
+        match console.poll() {
+            console_engine::events::Event::Key(KeyEvent { code, modifiers }) => {
+                if let Some(prompt) = w_login.get_active_prompt() {
+                    match code {
+                        KeyCode::Enter => {
+                            if w_login.is_prompting_username() {
+                                w_login.submit();
+                                console.print_fbg(
+                                    0,
+                                    0,
+                                    "                                      ",
+                                    Color::Red,
+                                    Color::Reset,
+                                );
+                                if w_login.get_username().is_empty() {
+                                    w_login.reset();
+                                    console.print_fbg(
+                                        0,
+                                        0,
+                                        "Username can't be empty!              ",
```

Comment:
You can do padding with `format!`, e.g. `format!("{:<17}", "Message")`, but here it would involve creating a string and referencing it in each loop (instead of a static str), so I'm not sure if (potentially) slightly better readability is worth it.